File: src/RuleEngine/RuleSet.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace RuleEngine
{
    public class RuleSet
    {
        public RuleSet(string name)
        {
            Name = name;
        }

        public string Name { get; set; }
        public IList<Rule> Rules { get; set; } = new List<Rule>();
        public IList<RuleOverride> RuleOverrides { get; set; } = new List<RuleOverride>();

        public RuleSet AddRule(Rule rule)
        {
            Rules.Add(rule);

            return this;
        }

        public RuleSet AddRuleOverride(RuleOverride ruleOverride)
        {
            RuleOverrides.Add(ruleOverride);

            return this;
        }
            
        public Proposition Evaluate(RuleContext context)
        {
            IDictionary<string, Proposition> ruleResults = new Dictionary<string, Proposition>();
            foreach (var r in Rules)
            {
                ruleResults.Add(r.Name, r.Evaluate(context));
            }

            foreach(var ro in RuleOverrides)
            {
                ruleResults.TryGetValue(ro.RuleName, out Proposition result);
                if (result != null)
                    result.Value = ro.Value;
            }

            bool finalResult = false;
            foreach(var pair in ruleResults)
            {
                finalResult = finalResult && (bool)pair.Value.Value;
            }
            return new Proposition(Name, finalResult);
        }
    }
}

File: src/RulesEngine.WebApp/Scripts/rules-engine/RuleSet.ts
import { Rule } from "./Rule"
import { RuleOverride } from "./RuleOverride"
import { Proposition } from "./Proposition"
import { RuleContext } from "./RuleContext"

export class RuleSet {
    constructor(name: string) {
        this.Name = name;
    }

    public Name: string = "";
    public Rules: Array<Rule> = [];
    public RuleOverrides: Array<RuleOverride> = [];

    public AddRule(rule: Rule): RuleSet {
        this.Rules.push(rule);
        return this;
    }

    public AddRuleOverride(ruleOverride: RuleOverride): RuleSet {
        this.RuleOverrides.push(ruleOverride);
        return this;
    }

    public Evaluate(context: RuleContext): Proposition {
        let ruleResults: Array<{ name: string, value: Proposition }> = []

        let count: number = this.Rules.length;
        for (let i: number = 0; i < count; i++) {
            let rule: Rule = this.Rules[i];
            ruleResults.push({ name: rule.Name, value: rule.Evaluate(context) });
        }

        count = this.RuleOverrides.length;
        let resultCount: number = ruleResults.length;
        for (let i: number = 0; i < count; i++) {
            let ruleOverride: RuleOverride = this.RuleOverrides[i];
            for (let j: number = 0; j < resultCount; j++) {
                let result: { name: string, value: Proposition } = ruleResults[j];
                if (result.name === ruleOverride.RuleName) {
                    ruleResults[j].value = new Proposition(ruleResults[j].value.Name, ruleOverride.Value);
                }
            }
        }

        let finalResult: boolean = false;
        for (let i: number = 0; i < resultCount; i++) {
            finalResult = finalResult && <boolean>ruleResults[i].value.Value;
        }

        return new Proposition(this.Name, finalResult);
    }
}